### PR TITLE
Added config for IBM Model M (ANSI/ISO) Teensy 2.0++ conversion

### PR DESCRIPTION
**Pull request by iw0rm3r**
## Description

This is my configuration for the reversible conversion of IBM's Model M keyboards (common 101-key ANSI and 102-key ISO models) from stock controller, to Teensy 2.0++ development board (produced by PJRC).

I am completely new to GitHub, but I tried to do everything right.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by iw0rm3r** [comment]
Hello and thank you for your comments!

> As this deals with converting an existing non-programmable keyboard to programmable through QMK, your files should be in `keyboards/converter/modelm101/` instead of `keyboards/modelm101`.

As I can see, `keyboards/converter/` directory is containing configurations for the converters of various interfaces (like ADB-to-USB), but my configuration is meant to be used for a replacement or original Model M controller, so no interface conversion is used here.
And, for example, directory 'keyboards/fc980c/' contains a configuration for replacement controller for Leopold board, which is exactly the same case.

> The `info.json` file deals with enabling the [QMK Configurator](https//config.qmk.fm/) to generate and compile keymaps for users who aren't as knowledgeable, comfortable or able to compile their firmware from source code.

Oh, I totally missed it!

>If you'd like help getting that set up, let me know. (I kind of specialize in it.)

Thanks! It would be great!

I've commited a wiring notes to README.md to clarify use of the config. Should i re-create a pull-request, or it will be automatically included?

**Comment by drashna** [comment]
To be honest, the leopold boards should probably be moved to the converter status, as well.

**Comment by iw0rm3r** [comment]
Ok, moved configuration to "converter" directory.

**Comment by noroadsleft** [comment]
> I've commited a wiring notes to README.md to clarify use of the config. Should i re-create a pull-request, or it will be automatically included?

As you may have already realized, if you open a pull request, GitHub tracks the fork and branch you're trying to have merged. Any additional commits that are pushed to that branch become part of the pull request as long as it remains open.


> > If you'd like help getting that set up, let me know. (I kind of specialize in it.)

> Thanks! It would be great!

The way QMK Configurator support works is like this:

Your keyboard's header file — specifically, https://github.com/qmk/qmk_firmware/blob/ab98024f6aa700333af71c5697a43d22025cd75a/keyboards/converter/modelm101/modelm101.h#L26-L34 dictates how many keycodes should be present on each layer of your keymap. Lines 26-45 here are what QMK calls a *layout macro* (remember this term).

`info.json` is a JSON (duh) file that describes the keyboard, and any layouts that are supported (the same PCB can support more than one layout, so the header file can describe more than one layout as well).

The best way to create this file is to go to [Keyboard Layout Editor](http://www.keyboard-layout-editor.com/) ("KLE") and recreate the physical layout that is described by your layout macro. When you're done, there should be the same number of keys described in your layout macro (Lines 26-34) as there are on the screen (KLE's Summary tab will tell you the key count). You should end up with something like this:

image

At this point, go to KLE's Raw Data tab, select all the text there, and copy it to your clipboard. Now, go to [QMK's KLE to info.json converter](https://qmk.fm/converter/), and paste the text you copied into the Input field. Click the Convert button, and after a moment, text will appear in the Output field. Select all of this text, copy it to the clipboard, then open your `info.json` file in a plain text editor (something like Wordpad, gedit, vim, or the like) and paste it. Save the file, stage and commit the changes, and push the commit. If all goes well, you should be able to load QMK Configurator after your pull request is merged, create a keymap there (export your keymap JSON so you can edit it later if needed; be advised this is a different from the `info.json` file you just made), compile it, save it and flash it to your controller, and it will work as expected.

There some are additional details [here](https://docs.qmk.fm/#/reference_configurator_support), and some technical details about the `info.json` format [here](https://docs.qmk.fm/#/reference_info_json) if you're interested.

If you have any issues, let me know.

**Comment by drashna** [comment]
Is this ready to go?

**Comment by iw0rm3r** [comment]
Thanks for the help on the JSON file!
I've just pushed it. Hope it's ok.

**Comment by drashna** [comment]
Well, it looks correct, at least.

**Comment by iw0rm3r** [comment]
Do I need to fix anything else?

File: keyboards/converter/modelm101/readme.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+Please note that some versions of Model M have LED connected with a ribbon cable, and not with separate wires.  
+
+Keyboard Maintainer: [iw0rm3r](https://github.com/iw0rm3r)  
+
+Hardware Supported: Teensy 2.0++ board by PJRC  
+Hardware Availability: https://www.pjrc.com/store/teensypp.html 
+
+Make example for this keyboard (after setting up your build environment):
+
+    make modelm101:default:teensy
```
Comment (noroadsleft) [inline keyboards/converter/modelm101/readme.md]: `make` command needs to match the directory tree under `keyboards/`; the files moved, so this needs to be updated.

```suggestion
    make converter/modelm101:default:teensy
```

File: keyboards/converter/modelm101/readme.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+Pins: F7 F6 F5 F4 F3 F2 F1 F0  
+--------------------------------------------------------  
+Status LEDs: CapsLock +5V ScrollLock NumLock  
+Pins:        B6       5V  B5         B4  
+```  
+
+Please note that some versions of Model M have LED connected with a ribbon cable, and not with separate wires.  
+
+Keyboard Maintainer: [iw0rm3r](https://github.com/iw0rm3r)  
+
```
Comment (noroadsleft) [inline keyboards/converter/modelm101/readme.md]: Ideally, this line should be deleted.

```suggestion
```

File: keyboards/converter/modelm101/modelm101.c [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "modelm101.h"
+
+void keyboard_pre_init_kb(void) {
+  /* Setting status LEDs pins to output and +5V (off) */
+  DDRB  |= (1<<4) | (1<<5) | (1<<6);
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]: Could you update these to use the GPIO commands, instead? 
https://docs.qmk.fm/#/internals_gpio_control

Eg, these should be:
```suggestion
  setPinOutput(B4);
  setPinOutput(B5);
  setPinOutput(B6);
```

File: keyboards/converter/modelm101/modelm101.c [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  PORTB |= (1<<4) | (1<<5) | (1<<6);
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]:
```suggestion
  writePinHigh(B4);
  writePinHigh(B5);
  writePinHigh(B6);
```

File: keyboards/converter/modelm101/modelm101.c [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+   * runs for every action, just before processing by the firmware */
+
+  return process_record_user(keycode, record);
+}
+
+void led_set_kb(uint8_t usb_led) {
+  if (usb_led & (1<<USB_LED_NUM_LOCK)) {
+    PORTB &= ~(1<<4);
+  } else {
+    PORTB |=  (1<<4);
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]:
```suggestion
    writePinHigh(B4);
 ```

File: keyboards/converter/modelm101/modelm101.c [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  }
+  if (usb_led & (1<<USB_LED_CAPS_LOCK)) {
+    PORTB &= ~(1<<6);
+  } else {
+    PORTB |=  (1<<6);
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]:
```suggestion
    writePinHigh(B5);
 ```

File: keyboards/converter/modelm101/modelm101.c [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  }
+  if (usb_led & (1<<USB_LED_SCROLL_LOCK)) {
+    PORTB &= ~(1<<5);
+  } else {
+    PORTB |=  (1<<5);
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]:
```suggestion
    writePinHigh(B5);
 ```

File: keyboards/converter/modelm101/modelm101.c [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  } else {
+    PORTB |=  (1<<4);
+  }
+  if (usb_led & (1<<USB_LED_CAPS_LOCK)) {
+    PORTB &= ~(1<<6);
+  } else {
+    PORTB |=  (1<<6);
+  }
+  if (usb_led & (1<<USB_LED_SCROLL_LOCK)) {
+    PORTB &= ~(1<<5);
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]:
```suggestion
    writePinLow(B5);
 ```

File: keyboards/converter/modelm101/modelm101.c [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
+  /* put your per-action keyboard code here
+   * runs for every action, just before processing by the firmware */
+
+  return process_record_user(keycode, record);
+}
+
+void led_set_kb(uint8_t usb_led) {
+  if (usb_led & (1<<USB_LED_NUM_LOCK)) {
+    PORTB &= ~(1<<4);
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]:
```suggestion
    writePinLow(B4);
 ```

File: keyboards/converter/modelm101/modelm101.c [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  /* put your per-action keyboard code here
+   * runs for every action, just before processing by the firmware */
+
+  return process_record_user(keycode, record);
+}
+
+void led_set_kb(uint8_t usb_led) {
+  if (usb_led & (1<<USB_LED_NUM_LOCK)) {
+    writePinLow(B4);
+ ```
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]: Whoops, I think this is my fault! 
```suggestion
```

File: keyboards/converter/modelm101/modelm101.c [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+  } else {
+    writePinHigh(B4);
+ ```
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]:
```suggestion
```

File: keyboards/converter/modelm101/modelm101.c [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  }
+  if (usb_led & (1<<USB_LED_CAPS_LOCK)) {
+    PORTB &= ~(1<<6);
+  } else {
+    PORTB |=  (1<<6);
+  }
+  if (usb_led & (1<<USB_LED_SCROLL_LOCK)) {
+    writePinLow(B5);
+ ```
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]:
```suggestion
```

File: keyboards/converter/modelm101/modelm101.c [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  } else {
+    writePinHigh(B5);
+ ```
```
Comment (drashna) [inline keyboards/converter/modelm101/modelm101.c]:
```suggestion
```

File: keyboards/converter/modelm101/modelm101.c (hunk 10 above)
Comment (iw0rm3r) [reply keyboards/converter/modelm101/modelm101.c]: No problem, I fixed this. Thanks for mentioning these functions, I didn't know about them.

File: keyboards/converter/modelm101/modelm101.c (hunk 6 above)
Comment (iw0rm3r) [reply keyboards/converter/modelm101/modelm101.c]: There was a misprint.

**Comment by iw0rm3r** [comment]
I replaced direct port commands. Is everything ok now?

**Comment by drashna** [comment]
Looks great!  Thanks!

**Pull request merged**